core/src/dird/ua_restore.cc

```diff
@@ -1183,13 +1204,40 @@ static bool BuildDirectoryTree(UaContext* ua, RestoreContext* rx)
       tree.FileCount = 0; /* set count to zero, no tree selection */
     }
   }
+}
 
+void FinishSelection(RestoreContext* rx, TreeContext& tree)
+{
+  // Walk down through the tree finding all files marked to be extracted
+  // making a bootstrap file.
+  for (TREE_NODE* node = FirstTreeNode(tree.root); node;
+       node = NextTreeNode(node)) {
+    Dmsg2(400, "FI=%d node=0x%x\n", node->FileIndex, node);
+    if (node->extract || node->extract_dir) {
+      Dmsg3(400, "JobId=%lld type=%d FI=%d\n", (uint64_t)node->JobId,
+            node->type, node->FileIndex);
+      /* TODO: optimize bsr insertion when jobid are non sorted */
+      AddDeltaListFindex(rx, node->delta_list);
+      AddFindex(rx->bsr.get(), node->JobId, node->FileIndex);
+      if (node->extract && node->type != TreeNodeType::NEWDIR) {
+        rx->selected_files++; /* count only saved files */
+      }
+    }
+  }
+}
+
+static bool SelectFiles(UaContext* ua,
+                        RestoreContext* rx,
+                        TreeContext& tree,
+                        bool done)
```

Comment:
```
/* We keep the tree with selected restore files.
   * For NDMP restores its used in the DMA to know what to restore.
   * The tree is freed by the DMA when its done. */
ua->jcr->dir_impl->restore_tree_root = tree.root;
```
This does not belong into this function.  I think we should handle this differently.
`TreeContext` should own its tree and its destructor should free it.  But there should be a function
`TreeContext::release_root() -> TREE_ROOT*/unique_ptr<TREE_ROOT>` or something like that by which one can take ownership of the root (TreeSelections root will then be set to nullptr).   That way if everything went right during `BuildDirectoryTree`, `SelectFiles`, etc. the caller can safely use
```
ua->jcr->dir_impl->restore_tree_root = ctx.release_root();
```
and on error the caller can leave the cleanup to the tree context.